lib/cnf_graph.ml

```diff
@@ -71,105 +71,182 @@ let pp_sat (fmt : Format.formatter) (res : Sat.res)
+            ])
+          tl
+        @ get_all_two_tuple_at_less_was_false tl
   in
-  let rec generate_exprs = function
+  let rec at_less_one_true = function
     | [] -> []
-    | hd :: tl when tl = [] -> only_var_true hd vars
-    | hd :: tl -> develop_or_cnf (only_var_true hd vars) (generate_exprs tl)
+    | hd :: tl -> (hd |> int_to_literal) :: at_less_one_true tl
   in
-  generate_exprs vars
-
-let check_has_color (x, y, t, c) (possible_colors : color list) (dim : int) :
-    cnf =
-  let rec aux res colors =
-    match colors with
-    | hd :: tl ->
-        [
-          (let var = get_name_variable (x, y, t, c) dim in
-           if hd = c then var |> int_to_literal
-           else var |> int_to_literal |> get_negation_of);
-        ]
-        :: aux res tl
-    | [] -> []
+  at_less_one_true vars :: get_all_two_tuple_at_less_was_false vars
+
+let check_each_case_has_only_one_color max_time w l
+    (possible_colors : color list) coef : cnf =
+  let rec get_all_cnf_each_case_has_only_one_color time width length : cnf =
+    if time < 0 then []
+    else if width < 0 then
+      get_all_cnf_each_case_has_only_one_color time (w - 1) (length - 1)
+    else if length < 0 then
+      get_all_cnf_each_case_has_only_one_color (time - 1) (w - 1) (l - 1)
+    else
+      let rec get_all_color colors =
+        match colors with
+        | hd :: tl ->
+            get_name_variable (width, length, time, hd) coef :: get_all_color tl
+        | [] -> []
+      in
+      get_only_one_true_cnf (get_all_color possible_colors)
+      @ get_all_cnf_each_case_has_only_one_color time (width - 1) length
   in
-  aux [] possible_colors
-
-let check_has_not_color (x, y, t, c) possible_colors dim : cnf =
-  [ get_name_variable (x, y, t, c) dim |> int_to_literal |> get_negation_of ]
-  :: get_only_one_true_cnf
-       (List.map
-          (fun color -> get_name_variable (x, y, t, color) dim)
-          possible_colors)
-
-let check_coloration_of_one_node (x : int) (y : int) (t : int)
-    (possible_colors : color list) (dim : int) : cnf =
-  let rec check_coord_have_one_color colors =
-    match colors with
-    | hd :: tl ->
-        develop_or_cnf
-          (check_has_not_color (x, y, t, hd) possible_colors dim)
-          (check_has_not_color ((x + 1) mod dim, y, t, hd) possible_colors dim
-          @ check_has_not_color (x, (y + 1) mod dim, t, hd) possible_colors dim
-          @ check_has_not_color
-              ((x + dim - 1) mod dim, y, t, hd)
-              possible_colors dim
-          @ check_has_not_color
-              (x, (y + dim - 1) mod dim, t, hd)
-              possible_colors dim)
-        @ check_coord_have_one_color tl
-    | [] -> []
+  get_all_cnf_each_case_has_only_one_color (max_time) (w - 1) (l - 1)
+
+let check_coloration_of_each_neighbor_is_different_for_each_node_of_the_graph w
+    l max_time possible_colors coef : cnf =
+  let check_coloration_of_each_neighbor_is_different_for_one_node (x : int)
```

Comment:
maybe you can split this function in few functions (check_coloration_of_each_neighbor_is_different_for_each_node_of_the_graph, check_coloration_of_each_neighbor_is_different_for_one_one, check_coord_have_one_color and check_coloration_of_each_node_each_time, because it's hard to read it